### PR TITLE
Make chuking smartly (long files) work on asr ctc_with_lm.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 1]
```diff
@@ -66,14 +66,27 @@ def ffmpeg_read(bpayload: bytes, sampling_rate: int) -> np.array:
     return audio
 
 
-def apply_stride(tokens, stride):
-    max_token_n = tokens.shape[-1]
+def audio_to_logits(tokens_or_logits, stride):
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/automatic_speech_recognition.py:69]: What exactly does the function do? Could we add some docstring? Also I don't understand the name `audio_to_logits` really

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 2]
```diff
@@ -266,7 +285,20 @@ def _forward(self, model_inputs):
...
+                logits = outputs["logits"].numpy()
+                stride = outputs.get("stride", None)
+                if stride is not None:
+                    total_n, left, right = stride
+                    # Total_n might be < logits.shape[1]
+                    # because of padding, that's why
+                    # we need to reconstruct this information
+                    # This won't work with left padding (which doesn't exist right now)
+                    right_n = total_n - right
+                    logits = logits[:, left:right_n]
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/automatic_speech_recognition.py:299]: Nice, yes I think that's the only approach that'll work right now

Comment (patrickvonplaten) [reply src/transformers/pipelines/automatic_speech_recognition.py:299]: Padding is not possible here sadly

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:69]: It's just change the stride number to go from `audio space` (10s at 16_000  means (160_000, 8_000, 8_000)) for instance. to `logits_space` (2333, 160, 160) for instance.

Do you think of a better name ? Doctstring could help a little.

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 2 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:299]: Padding works, it's handled by the batching mecanism. All I mentionned here is why we need so much information (`logits` might get padded while `stride` no).

We *can* make it work, relatively trivially for left padding.

```python
if self.tokenizer.padding_side == "left":
    left = logits.shape[1] - total_n + left_n
    right = logits.shaoe[1] - total_n + right_n
else:
   left = left_n
   right_n = total_n - right_n
```
Just thought this was overly complex since left padding doesn't seem likely here.

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 3]
```diff
@@ -66,14 +66,36 @@ def ffmpeg_read(bpayload: bytes, sampling_rate: int) -> np.array:
     return audio
 
 
-def apply_stride(tokens, stride):
-    max_token_n = tokens.shape[-1]
+def audio_to_logits(tokens_or_logits, stride):
+    # Shape is [B, SEQ] for tokens
+    # [B, SEQ, V] for logits
+
+    is_batch_size_one = isinstance(stride, tuple) and not isinstance(stride[0], tuple)
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/automatic_speech_recognition.py:73]: added this hack to make it work with batch size 1. Maybe we can make this function a bit "prettier"? Or just give it a better docstring?

File: tests/test_pipelines_automatic_speech_recognition.py [hunk 4]
```diff
@@ -295,6 +295,37 @@ def test_chunking_fast(self):
...
+        ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation").sort("id")
+        audio = ds[40]["audio"]["array"]
+
+        n_repeats = 2
+        audio_tiled = np.tile(audio, n_repeats)
+        output = speech_recognizer([audio_tiled], batch_size=2)
+        self.assertEqual(output, [{"text": ANY(str)}])
+        self.assertEqual(output[0]["text"][:6], "<s> <s")
+
+    def test_chunking_fast_with_lm_batch_size_1(self):
```
Comment (patrickvonplaten) [inline tests/test_pipelines_automatic_speech_recognition.py:314]: this test failed before @Narsil - added it here for you to check. Happy to remove it or rename it afterwards

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/pipelines/automatic_speech_recognition.py:308]: Ah I see - ok yeah

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/pipelines/automatic_speech_recognition.py:69]: Ah I see. Maybe `get_output_stride_from_input_stride(input_shape, stride)` and directly pass the shape? Yeah think a little docstring can help here

**Review by patrickvonplaten — APPROVED** [review]
Super cool! Works well for the swedish example of:

```bash
./eval.py --model_id hf-test/xls-r-300m-sv --dataset speech-recognition-community-internal/tedx_manual_dev_test --config sv --split validation --chunk_length_s 5.0 --stride_length_s 1.0
```

with the eval script here: https://github.com/huggingface/transformers/blob/master/examples/research_projects/robust-speech-event/eval.py

Made some small changes to make it work for batch size 1 - feel free to refactor those to make it cleaner @Narsil

File: tests/test_pipelines_automatic_speech_recognition.py (hunk 4 above)
Comment (Narsil) [reply tests/test_pipelines_automatic_speech_recognition.py]: Kept the test, but fused it with the previous one.
